Postgresql Qgis Project versioning

JanCaha opened the pull request:
## Description

This allows storing older versions of QGIS projects in PostgreSQL. The process is mostly handled by a function and a trigger in the DB. 

There are two places to handle the turning on versions of projects and loading previous versions.

### Browser

On schema, user has option to enable the feature (with confirmation dialog). The action is disabled if the versioning is already active for the schema.
<img width="519" height="218" alt="2025-10-27-11-44-03" src="https://github.com/user-attachments/assets/bdb7b698-7314-4960-ad18-64ffbb5d645c" />

On project user either has option to allow versioning on the parent schema of the project 

<img width="298" height="234" alt="2025-10-27-11-45-27" src="https://github.com/user-attachments/assets/3213e749-1e30-4683-b531-d725af1cbfd2" />

If the versioning is already active the user can see the versions of the project.

<img width="273" height="214" alt="2025-10-27-11-45-13" src="https://github.com/user-attachments/assets/00dd71f4-369b-455a-8c26-bdf1a06c1956" />

<img width="600" height="276" alt="versions jpg" src="https://github.com/user-attachments/assets/caca6400-ea95-4388-b948-c6b60fdb5466" />

### Open From.../Save To... 

In those menus the user can also enable the feature and select previous version to load.

<img width="552" height="474" alt="save_project" src="https://github.com/user-attachments/assets/88c856b7-1b04-4565-ba4c-6b006f38e0e8" />

When loading QGIS project from schema that has versions activated the user can select older version of the project to load, by selecting it in the QTableWidget. If nothing is selected there the active version of QGIS project is loaded.

<img width="552" height="474" alt="load_project" src="https://github.com/user-attachments/assets/7761788e-21e2-4cbb-bcd9-2524097c7193" />

Fixes #62299

Funded by: Ocean Winds

lbartoletti committed:
feat(Transect): Add option to generate a transect at a fixed distance (ignore existing vertices)

lbartoletti committed:
feat(processing): add separate transect fixed distance algorithm

lbartoletti committed:
test(transect): Add a test for transect with fixed distance

JanCaha committed:
new utils functions

JanCaha committed:
add info to QgsPostgresProjectUri

JanCaha committed:
work with QgsPostgresProjectUri parameters

JanCaha committed:
if project uri specifies version load from backup, otherwise load project normally

JanCaha committed:
update store dialog

JanCaha committed:
update to QgsScrollArea

JanCaha committed:
remove versions widget, it is handled in the code

JanCaha committed:
table widget

JanCaha committed:
QAbstractTableModel for PG QGIS project versions

JanCaha committed:
selection dialog for PG QGIS project versions

JanCaha committed:
use model to handle versions

JanCaha committed:
store info about versions of qgis project in QgsPGSchemaItem

JanCaha committed:
project version in browser

JanCaha committed:
Merge branch 'master' into postgresql-qgis-project-versioning

JanCaha committed:
change the structure to allow for default values

JanCaha committed:
cast to int

BathoryPeter commented:
Thank you for this great feature!

If I understand correctly, you only keep the 10 latest versions, and this number is hard-coded into the trigger. It would be worth making this configurable at the project level.

It would also be useful to be able to select versions to keep. This can be done with a checkbox, or even by naming the version in the version management dialog.

JanCaha commented:
Not yet :-D There are still few edge cases to think about...

JanCaha committed:
Merge branch 'master' into postgresql-qgis-project-versioning

JanCaha committed:
function to move project versions to another schema

JanCaha committed:
when moving project move versions as well if it is enabled

lbartoletti committed:
fix(processing): segmentize curved geometries and use safe casting in transect algorithms

lbartoletti committed:
refactor(processing): eliminate unnecessary QgsLineString copy in transect base

lbartoletti committed:
refactor(processing): use switch statement for transect orientation logic

lbartoletti committed:
refactor(processing): use QgsProcessingParameterDistance for transect interval

lbartoletti committed:
perf(processing): use visitPointsByRegularDistance for fixed distance transects

lbartoletti committed:
feat(processing): set Both as default side for transect algorithms

lbartoletti committed:
refactor(processing): improve transect parameter ordering

lbartoletti committed:
feat(processing): add option to include transect at start of line

lbartoletti committed:
fix(processing): correctly add start point when include option is enabled

rouault committed:
QgsGml: handle implicit srsDimension=3 when a 3D SRS is specified through srsName

rouault committed:
[OAPIF]: handle mResponseHeaders when sendGET() is sent multiple times on the same object

rouault committed:
[OAPIF] Add support for GML as a feature format

Add support for GML. Users may select it in the combo box
in the "Modify WFS Connection" dialog if the server
exposes it as a feature format.

Several scenarios are supported:
- with or without XML schema available through a rel=describedby
  link.  As WFS, the XML schema can be examined with QGIS simple
  schema parser, or with GDAL GMLAS one, depending on the selected "Feature mode"
- with or without bulk download link available through a rel=enclosure link

rouault committed:
[WFS provider] Move code related to XML Schema analysis to a new helper class and QgsOgcUtils, so it can be reused by OAPIF

lbartoletti committed:
refactor(processing): add QgsAlgorithmTransectBase used for fixed distance and 'normal' one

lbartoletti committed:
fix(processing): validate closestSegment result and prevent negative vertex index in transect algorithm

lbartoletti committed:
chore: pet clang-format

alexbruy committed:
expose output nodata and raster creation options in the native Aspect
algorithm

DelazJ committed:
Remove unused help text, from algorithms already ported to c++

lbartoletti committed:
build: remove processAlgorithm from qgsalgorithmtransect.h

DelazJ committed:
Add Python-based algorithms short help and description to their .py file

DelazJ committed:
Cleanup references to the now unnecessary qgis.yaml file

lbartoletti committed:
chore(transect): fix duplicate from previous rebase

rouault committed:
WFS/OAPIF connection dialog: enable 'force initial getfeature' checkbox only for WFS, and 'feature mode' only for GML

rouault committed:
Re-order inclusions

nyalldawson committed:
Fix crash when drawing layout map grids

nyalldawson committed:
Add QgsAlignmentComboBox API to set/retrieve alignment using QGIS enums

alexbruy committed:
expose output nodata and raster creation options in the native Slope
algorithm

alexbruy committed:
expose output nodata and raster creation options in the native Hillshade
algorithm

alexbruy committed:
iexpose output nodata and raster creation options in the native
Ruggedness index algorithm

alexbruy committed:
add native total curvature algorithm from the terrain analysis library

nyalldawson committed:
Promote map grid enums to enum class, move to Qgis

nyalldawson committed:
Fix failing pre-commit check

nyalldawson committed:
Remove unused orphan files

nyalldawson committed:
Fix use of banned keywords

nyalldawson committed:
Rename test class to match naming convention

nyalldawson committed:
Fix extraction of non-standard class names

nyalldawson committed:
QString fixup

nyalldawson committed:
Fix crash in data defined assistants when min value > max value

basilrabi committed:
[RPM] modernize specfile

Use cmake and pkgconfig in BuildRequires

paleolimbot committed:
add array stream wrapper

basilrabi committed:
[RPM] enable QgsMagneticModel

paleolimbot committed:
add stream exporter

paleolimbot committed:
python

alexbruy committed:
make algorithm more thread safe by retrieving layer source in the
prepareAlgorithm() step

nyalldawson committed:
Apply suggestions from code review

basilrabi committed:
fix PDF4QT build with blend2d >= 0.20

Since
https://github.com/blend2d/blend2d/commit/320a63aaf11c028a650a6b7d54cd3b7b12d8feed ,
the code base was refactored to use snake_case convention. This change
is needed to compile PDF4QT in distros with updated blend2d package.

DelazJ committed:
Adjust help texts

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

alexbruy committed:
retrieve layer source in the prepareAlgorithm() step

JanCaha committed:
Merge branch 'master' into postgresql-qgis-project-versioning

JanCaha committed:
fix precommit issue

JanCaha committed:
fix precommit issues

JanCaha committed:
more precommit fixes

alexbruy committed:
fix layout in the Legend tab of the vector layer properties dialog (fix #64319)

rouault committed:
QgsGeometry::coerceToType(): make it handle geometry collection of point/line/polygon to multipoint/line/polygon

rouault committed:
WFS/OAPIF: use gsGeometry::coerceToType()

DelazJ committed:
Fix line length

paleolimbot committed:
more const methods

paleolimbot committed:
format

paleolimbot committed:
fix exception handling

Commit:
auto sipify 🍺

paleolimbot committed:
add asserts for the size of unsigned long long

paleolimbot committed:
tweak comment to rerun ci

Commit:
auto sipify 🍺

nyalldawson committed:
Fix test

nyalldawson committed:
Make code_fixup script a pre-commit script

nyalldawson committed:
Remove unused chkspelling script

nyalldawson committed:
Move fix_allows_to.sh logic to prepare_commit_code_fixup.py

nyalldawson committed:
Fix grammar

nyalldawson committed:
Move more pre-commit dependancies to pre_commit dir

nyalldawson committed:
Fix cppcheck warning

nyalldawson committed:
Fix build

Commit:
auto sipify 🍺

Commit:
auto sipify 🍺

JanCaha committed:
Algorithm to create Cloud Optimized GeoTiff (#63911)

uclaros commented on src/providers/postgres/qgspostgresdataitems.h:
```diff
@@ -76,10 +76,25 @@ class QgsPGSchemaItem : public QgsDatabaseSchemaItem
+     *
+     * \since QGIS 4.0
+     */
+    void setProjectVersioning( const bool active ) { mProjectVersioningActive = active; }
```
let's go with `setProjectVersioningEnabled`/`projectVersioningEnabled`

uclaros reviewed (commented):
Nice!
Just a couple of comments.
Also, can you add a vertical spacer at the bottom of the `open from` dialog so the table of versions is not 'floating'?

JanCaha committed:
rename variable

JanCaha committed:
rename functions

JanCaha committed:
move notification to proper place

JanCaha committed:
rename function and return if it was successful

JanCaha committed:
rebase
